feat(all): add cacheable queries to the magento cacheable queries list

lderrickable opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the new behavior?
Cacheable queries are now cacheable by Magento through the apollo link generators.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

damienwebdev commented on apps/demo/src/app/drivers/magento/magento.module.ts:
```diff
@@ -39,12 +45,13 @@ export class DemoMagentoDriverModule {
           if (networkError) console.log(`[Network error]: ${networkError}`);
-        }),
+				}),
+				this.apolloCacheableOperationsLinkGenerator.getLink(),
```
variable name should probably be -- magtentoLinkGen.getLink()

lderrickable replied:
changed

damienwebdev reviewed (changes requested):
Take a gander at https://devdocs.magento.com/guides/v2.4/graphql/caching.html -- I don't think countries are cachable.

lderrickable commented:
> Take a gander at https://devdocs.magento.com/guides/v2.4/graphql/caching.html -- I don't think countries are cachable.

fixed. Also, I removed the provider functions. It fails the build because of the "no function calls in decorators" aot rule.

damienwebdev commented:
@lderrickable the provider functions can be kept, but they need to be written like:

```
export function myFunction(): {
 return "test";
}

not 

export const myFunction = () => {
return "test";
};
```

lderrickable commented:
@damienwebdev I don't think that works. Seems like the point of restricting function calls in decorators is to ensure that you can't pass in data at runtime. I think the only reason you can do the factory function thing is because it forces you not to be able to pass anything in. Their example is:

```
export function serverFactory() {
  return new Server();
}

@Component({
  ...
  providers: [{provide: server, useFactory: serverFactory}]
})
```

I think the only reason you're allowed to do this is because you can't pass anything into the function, so even though it is a function, it can be completely understood before runtime. I think the only way we could make this work is something like:

```
export function navigationCacheableQueriesProvider() {
  return ({
     provide: MAGENTO_OPERATIONS_TOKEN,
     useValue: NAVIGATION_QUERY_NAME,
     multi: true
  })
}

@Component({
  ...
  providers: [{provide: server, useFactory: navigationCacheableQueriesProvider}]
})
```

I don't even know if that would work, but it certainly defeats the purpose of having the functions anyway. Maybe I'm wrong about this, but that's what I gathered from the angular docs. I tried using the new function you wrote too. It doesn't work

lderrickable commented:
@damienwebdev Added the provider functions for angular 9+ apps

griest024 commented on libs/driver/magento/src/graphql/cacheable-operations/cacheable-operations-token.ts:
```diff
@@ -3,6 +3,29 @@ import { InjectionToken } from '@angular/core';
+/**
+ * Adds many operation names to the list of cacheable Magento operations. Use only with Angular 9+.
+ */
+export function provideManyDaffMagentoCacheableOperations(operationNames: string[]) {
```
I think its a little cleaner to have this be a rest param.

lderrickable replied:
I'm not sure what you mean

griest024 replied:
Like `provideManyDaffMagentoCacheableOperations(...operationNames: string[])` so you can just pass args into the function instead of an array.

lderrickable replied:
oh, I didn't know that's what that was called. Changed
